code/parse/sexp.cpp

```diff
@@ -1776,6 +1776,120 @@ int count_items_with_name(const char *name, const T* item_array, int num_items)
 	return count;
 }
 
+// helper functions for check_container_value_data_type()
+bool check_container_data_sexp_arg_type(ContainerType con_type, bool is_string, bool is_number)
+{
+	if (any(con_type & ContainerType::STRING_DATA)) {
+		return is_string;
+	} else if (any(con_type & ContainerType::NUMBER_DATA)) {
+		return is_number;
+	} else {
+		UNREACHABLE("Unknown container data type %d", (int)con_type);
+		return false;
+	}
+}
+
+bool check_map_container_key_sexp_arg_type(ContainerType con_type, bool is_string, bool is_number)
+{
+	if (any(con_type & ContainerType::STRING_KEYS)) {
+		return is_string;
+	} else if (any(con_type & ContainerType::NUMBER_KEYS)) {
+		return is_number;
+	} else {
+		UNREACHABLE("Unknown map container key type %d", (int)con_type);
+		return false;
+	}
+}
+
+// Checks "container value" SEXP arg type
+// follows return value conventions of check_sexp_syntax()
+int check_container_value_data_type(int op, int argnum, ContainerType con_type, bool is_string, bool is_number)
+{
+	switch (op) {
+		case OP_LIST_HAS_DATA:
+		case OP_LIST_DATA_INDEX:
+		case OP_MAP_HAS_DATA_ITEM:
+		case OP_CONTAINER_ADD_TO_LIST:
+		case OP_CONTAINER_REMOVE_FROM_LIST:
+			if (!check_container_data_sexp_arg_type(con_type, is_string, is_number)) {
+				return SEXP_CHECK_WRONG_CONTAINER_DATA_TYPE;
+			}
+			break;
+
+		case OP_MAP_HAS_KEY:
+		case OP_CONTAINER_REMOVE_FROM_MAP:
+			Assert(any(con_type & ContainerType::MAP));
+			if (!check_map_container_key_sexp_arg_type(con_type, is_string, is_number)) {
+				return SEXP_CHECK_WRONG_MAP_KEY_TYPE;
+			}
+			break;
+
+		case OP_CONTAINER_ADD_TO_MAP:
+			Assert(any(con_type & ContainerType::MAP));
+			if (argnum % 2 != 0) {
```

Comment:
Shouldn't this be == 0?  Don't keys come before data?  Arg0 - key, Arg1 - data, etc.

Reply:
Since `add--to-map`'s first argument is the name of a map container, the SEXP's map keys start at `argnum` 1 and values start at `argnum` 2.